File: src/devtwin/system/docker.py
"""Safe, read-only Docker inspection.

DevTwin never stops, removes, or otherwise mutates containers. Every
function here maps to a read-only ``docker`` subcommand.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path

from devtwin.core.models import DictMixin, Presence
from devtwin.core.runner import run_command, which


@dataclass
class ContainerInfo(DictMixin):
    name: str
    image: str
    status: str
    ports: str = ""


@dataclass
class DockerInfo(DictMixin):
    presence: Presence
    version: str | None = None
    daemon_running: bool = False
    containers: list[ContainerInfo] = field(default_factory=list)
    compose_files: list[str] = field(default_factory=list)


COMPOSE_FILE_NAMES = [
    "compose.yaml",
    "compose.yml",
    "docker-compose.yaml",
    "docker-compose.yml",
]


def find_compose_files(root: Path) -> list[str]:
    return [name for name in COMPOSE_FILE_NAMES if (root / name).exists()]


def inspect_docker(root: Path | None = None) -> DockerInfo:
    if which("docker") is None:
        return DockerInfo(
            presence=Presence.NOT_INSTALLED,
            compose_files=find_compose_files(root) if root else [],
        )

    version_result = run_command(
        ["docker", "version", "--format", "{{.Server.Version}}"], timeout=5
    )
    daemon_running = version_result.available and version_result.returncode == 0
    version_str: str | None = None
    if daemon_running:
        version_str = version_result.stdout.strip() or None
    else:
        # Daemon down, but CLI installed -- try the client-only version.
        client_version = run_command(["docker", "--version"], timeout=5)
        if client_version.available:
            version_str = client_version.stdout.strip() or None

    containers: list[ContainerInfo] = []
    if daemon_running:
        ps_result = run_command(
            ["docker", "ps", "--format", "{{json .}}"],
            timeout=8,
        )
        if ps_result.available and ps_result.returncode == 0:
            for line in ps_result.stdout.splitlines():
                line = line.strip()
                if not line:
                    continue
                try:
                    raw = json.loads(line)
                except json.JSONDecodeError:
                    continue
                containers.append(
                    ContainerInfo(
                        name=raw.get("Names", ""),
                        image=raw.get("Image", ""),
                        status=raw.get("Status", ""),
                        ports=raw.get("Ports", ""),
                    )
                )

    return DockerInfo(
        presence=Presence.INSTALLED,
        version=version_str,
        daemon_running=daemon_running,
        containers=containers,
        compose_files=find_compose_files(root) if root else [],
    )

File: mcphub/devtwin/src/system/docker.ts
/**
 * Safe, read-only Docker inspection.
 *
 * DevTwin never stops, removes, or otherwise mutates containers. Every
 * function here maps to a read-only `docker` subcommand.
 */

import { Presence } from '../core/models.js';
import { runCommand, which } from '../core/runner.js';
import { existsAny } from './filesystem.js';

export interface ContainerInfo {
  name: string;
  image: string;
  status: string;
  ports: string;
}

export interface DockerInfo {
  presence: Presence;
  version: string | null;
  daemon_running: boolean;
  containers: ContainerInfo[];
  compose_files: string[];
}

export const COMPOSE_FILE_NAMES = [
  'compose.yaml',
  'compose.yml',
  'docker-compose.yaml',
  'docker-compose.yml',
];

export function findComposeFiles(root: string): string[] {
  return existsAny(root, COMPOSE_FILE_NAMES);
}

export async function inspectDocker(root?: string | null): Promise<DockerInfo> {
  if (which('docker') === null) {
    return {
      presence: Presence.NOT_INSTALLED,
      version: null,
      daemon_running: false,
      containers: [],
      compose_files: root ? findComposeFiles(root) : [],
    };
  }

  const versionResult = await runCommand(['docker', 'version', '--format', '{{.Server.Version}}'], {
    timeout: 5,
  });
  const daemonRunning = versionResult.available && versionResult.returncode === 0;
  let versionStr: string | null = null;
  if (daemonRunning) {
    versionStr = versionResult.stdout.trim() || null;
  } else {
    // Daemon down, but CLI installed -- try the client-only version.
    const clientVersion = await runCommand(['docker', '--version'], { timeout: 5 });
    if (clientVersion.available) {
      versionStr = clientVersion.stdout.trim() || null;
    }
  }

  const containers: ContainerInfo[] = [];
  if (daemonRunning) {
    const psResult = await runCommand(['docker', 'ps', '--format', '{{json .}}'], { timeout: 8 });
    if (psResult.available && psResult.returncode === 0) {
      for (const rawLine of psResult.stdout.split(/\r?\n/)) {
        const line = rawLine.trim();
        if (!line) {
          continue;
        }
        let raw: Record<string, unknown>;
        try {
          raw = JSON.parse(line) as Record<string, unknown>;
        } catch {
          continue;
        }
        containers.push({
          name: asString(raw['Names']),
          image: asString(raw['Image']),
          status: asString(raw['Status']),
          ports: asString(raw['Ports']),
        });
      }
    }
  }

  return {
    presence: Presence.INSTALLED,
    version: versionStr,
    daemon_running: daemonRunning,
    containers,
    compose_files: root ? findComposeFiles(root) : [],
  };
}

/** `dict.get(key, "")` equivalent that also tolerates non-string JSON values. */
function asString(value: unknown): string {
  return typeof value === 'string' ? value : '';
}
